File: runtimes-demo/vertx/src/main/kotlin/demo/runtimes/vertx/Greetings.kt
package demo.runtimes.vertx

import io.vertx.core.http.HttpServerOptions
import io.vertx.ext.web.Router
import io.vertx.ext.web.RoutingContext
import io.vertx.ext.web.client.WebClient
import io.vertx.kotlin.coroutines.CoroutineVerticle
import io.vertx.micrometer.PrometheusScrapingHandler

class Greetings(private val client: WebClient) : CoroutineVerticle() {
  private val peers = listOf("quarkus", "nodejs", "springboot")

  override suspend fun start() {
    // Register stadium API
    val serverOptions = HttpServerOptions().setPort(8080)
    val router = Router.router(vertx)
    router.route("/metrics").handler(PrometheusScrapingHandler.create())

    router["/health"].handler { it.response().end() }
    router["/greetings"].handler { greetings(it) }

    vertx.createHttpServer().requestHandler(router)
            .listen(serverOptions.port, serverOptions.host)
  }

  private fun greetings(ctx: RoutingContext) {
    val from = ctx.request().getParam("from")
    if (from == null) {
      ctx.response().end("Hello, I'm Vert.X!")
      peers.forEach { peer ->
        client.get(8080, peer, "/greetings?from=vertx").send {
          if (it.succeeded()) {
            println(it.result().bodyAsString())
          } else {
            it.cause().printStackTrace()
          }
        }
      }
    } else {
      val resp = "Hello $from, I'm Vert.X!"
      println(resp)
      ctx.response().end(resp)
    }
  }
}

File: runtimes-demo/nodejs/greetings.js
const express = require('express');
const http = require('http');
const { collectDefaultMetrics, register } = require('prom-client');

collectDefaultMetrics();
const app = express();

app.use('/greetings', (request, response) => {
  const from = request.query.from;
  if (!from) {
    response.send("Hello, I'm NodeJS!");
    ['vertx', 'quarkus', 'springboot'].forEach(peer => {
      console.log('client call to ' + peer);
      http.get(`http://${peer}:8080/greetings?from=nodejs`, {}, res => {
        let str = ''
        res.on('data', chunk => str += chunk);
        res.on('end', () => {
          console.log(str);
        });
      }).on('error', e => {
        console.log("Got error: " + e.message);
      });
    });
  } else {
    const resp = (`Hello ${from}, I'm NodeJS!`);
    console.log(resp);
    response.send(resp);
  }
});

// expose our metrics at the default URL for Prometheus
app.get('/metrics', (_, response) => register.metrics().then(metrics => {
  response.set('Content-Type', register.contentType);
  response.send(metrics);
}));

app.listen(8080, () => console.log(`Greetings app listening on port 8080!`));
